Utils/XDrive.hpp

```diff
@@ -57,8 +57,130 @@ class XDriveRobot {
         frontrightdrive.stop(bt);
     }
 
+    void resetRot() {
+        backleftdrive.resetRotation();
+        backrightdrive.resetRotation();
+        frontleftdrive.resetRotation();
+        frontrightdrive.resetRotation();
+    }
+
+    double frxdriveVelocity (double speed, double tgtHeading, double wheelAngle = 45) {
+        return speed * sin(degree2arc(90-wheelAngle-tgtHeading));
+    }
+
+    double flxdriveVelocity (double speed, double tgtHeading, double wheelAngle = 45) {
+        return speed * sin(degree2arc(90-wheelAngle+tgtHeading));
+    }
+
+    double brxdriveVelocity (double speed, double tgtHeading, double wheelAngle = 45) {
+        return speed * sin(degree2arc(90-wheelAngle+tgtHeading));
+    }
+
+    double blxdriveVelocity (double speed, double tgtHeading, double wheelAngle = 45) {
+        return speed * sin(degree2arc(90-wheelAngle-tgtHeading));
+    }
+
+    Coord goStraight(double dist, vex::directionType dt, double tgtHeading, double originalSpeed, double kp = 0.02, vex::brakeType bt = brake, Coord srcLoc = Coord(0.0, 0.0));
+
 };  // class XDriveRobot
 
+Coord XDriveRobot::goStraight(double dist, vex::directionType dt, double tgtHeading, double originalSpeed, double kp, vex::brakeType bt, Coord srcLoc) {
+  Coord currLoc = srcLoc;
+
+  resetRot();
+
+  // orientation of the wheels, default = 45
+  double wheelAngle = 45;
+  double wheelArc = degree2arc(45);
+
+  // target orientation that the robot should cont. to travel at
+  // tgtHeading is a parameter
+  double targetArc = degree2arc(tgtHeading);
+  
+  // tracking distance
+  double distToGo = dist; // distance more to travel
+  double distTravelled = dist - distToGo; // distance already traveled
+  double prevRot = backleftdrive.rotation(vex::rotationUnits::deg); // amount of rotation at the last run through
+  double changedRot = backleftdrive.rotation(vex::rotationUnits::deg) - prevRot; // rotations passed since last loop
+
+  // speed
+  double finalSpeed = 10; // capping speed
+  double speed = originalSpeed; // max speed
+  double const adaptiveInterval = 10; // slow down/speed up interval
+
+  
+  // location tracking on the coordinate plane
+  double dx = 0; // change in x coordinate since last loop
+  double dy = 0; // change in y coordinate since last loop
+
+  double xSpeed = originalSpeed * cos(targetArc);
+  double ySpeed = originalSpeed * sin(targetArc);
+
+  double blspeed = xSpeed - ySpeed;
+  double brspeed = -xSpeed - ySpeed;
+  double flspeed = -xSpeed + ySpeed;
+  double frspeed = xSpeed + ySpeed;
+
+  while (distToGo > 0) {
+
+    double frsign = 1;
```

Comment:
pls move such assignments outside of loop as apparently they are loop invariant.